Mathlib/CategoryTheory/Limits/Shapes/Pullback/Categorical/Basic.lean

```diff
@@ -0,0 +1,354 @@
+`CategoricalPullback F G` as the category of triples
+`(a : A, c : C, e : F.obj a ≅ G.obj b)`.
+
+The category `CategoricalPullback F G` sits in a canonical `CatCommSq`, and we formalize that
+this square is a "limit" in the following sense: functors `X ⥤ CategoricalPullback F G` are
+equivalent to pairs of functors `(L : X ⥤ A, R : X ⥤ C)` equipped with a natural isomorphism
+`L ⋙ F ≅ R ⋙ G`.
+
+We formalize this by introducing a category `CatCommSqOver F G X` that encodes
+exactly this data, and we prove that the category of functors `X ⥤ CategoricalPullback F G` is
+equivalent to `CatCommSqOver F G X`.
+
+## Main declarations
+
+* `CategoricalPullback F G`: the type of the categorical pullback.
+* `πₗ F G : CategoricalPullback F G` and `πᵣ F G : CategoricalPullback F G`: the canonical
+  projections.
+* `CategoricalPullback.catCommSq`: the canonical `CatCommSq (πₗ F G) (πᵣ F G) F G` which exhibits
+  `CategoricalPullback F G` as the pullback (in the (2,1)-categorical sense)
+  of the cospan of `F` and `G`.
+* `CategoricalPullback.functorEquiv F G X`: the equivalence of categories between functors
+  `X ⥤ CategoricalPullback F G` and `CatCommSqOver F G X`, where the latter is an abbrev for
+  `CategoricalPullback (whiskeringRight X A B|>.obj F) (whiskeringRight X C B|>.obj G)`.
+
+## References
+* [Kerodon: section 1.4.5.2](https://kerodon.net/tag/032Y)
+* [Niles Johnson, Donald Yau, *2-Dimensional Categories*](https://arxiv.org/abs/2002.06055),
+  example 5.3.9, although we take a slightly different (equivalent) model of the object.
+
+## TODOs:
+* 2-functoriality of the construction with respect to "transformation of categorical
+  cospans".
+* Full equivalence-invariance of the notion (follows from suitable 2-functoriality).
+* Define a `CatPullbackSquare` typeclass extending `CatCommSq`that encodes the
+  fact that a given `CatCommSq` defines an equivalence between the top left
+  corner and the categorical pullback of its legs.
+* Define a `IsCatPullbackSquare` propclass.
+* Define the "categorical fiber" of a functor at an object of the target category.
+* Pasting calculus for categorical pullback squares.
+* Categorical pullback squares attached to Grothendieck constructions of pseudofunctors.
+* Stability of (co)fibered categories under categorical pullbacks.
+
+-/
+
+universe v₁ v₂ v₃ v₄ u₁ u₂ u₃ u₄
+
+namespace CategoryTheory.Limits
+
+attribute [local simp] CatCommSq.iso_hom_naturality  CatCommSq.iso_inv_naturality
+
+section
+
+variable {A : Type u₁} {B : Type u₂} {C : Type u₃}
+  [Category.{v₁} A] [Category.{v₂} B] [Category.{v₃} C]
+  (F : A ⥤ B) (G : C ⥤ B)
+
+/-- The `CategoricalPullback F G` is the category of triples
+`(a : A, c : C, F a ≅ G c)`.
+Morphisms `(a, c, e) ⟶ (a', c', e')` are pairs of morphisms
+`(f₁ : a ⟶ a', f₂ : c ⟶ c')` compatible with the specified
+isomorphisms. -/
+@[kerodon 032Z]
+structure CategoricalPullback where
+  /-- the left element -/
+  left : A
```

Comment:
For pullback/binary products, we tend to use `fst/snd` rather than `left/right` (which appear in pushouts as `inl/inr`), so that `fst/snd` could be slightly better names than `left/right`?